### DISCUSSION TITLE
CSS theme color variables

### DISCUSSION THREAD

**Discussion by techniq** [tailwindlabs/tailwindcss#6137] [Ideas]
## Current

I find myself needing to access the Tailwind color palette as CSS variables often, particularly when working with components in Svelte/Vue/React.

I currently use a plugin to expose the entire palette on `:root`...

```js
// Expose color palette as CSS variables (--color-xxx-yyy) - https://gist.github.com/Merott/d2a19b32db07565e94f10d13d11a8574
plugin(function ({ addBase, theme }) {
  function extractColorVars(colorObj, colorGroup = '') {
    return Object.keys(colorObj).reduce((vars, colorKey) => {
      const value = colorObj[colorKey];

      const newVars =
        typeof value === 'string'
          ? { [`--color${colorGroup}-${colorKey}`]: value }
          : extractColorVars(value, `-${colorKey}`);

      return { ...vars, ...newVars };
    }, {});
  }

  addBase({
    ':root': extractColorVars(theme('colors')),
  });
}),
```
but this has drawbacks:
- pollutes the CSS variable namespace (since variables are all inherited)
- Increases bundle size since these values are not JIT'd
- Requires a plugin

![image](https://user-images.githubusercontent.com/177476/142649337-68ffaa2c-3d32-4232-9417-c2aa264a9996.png)

While you can currently access the theme within a `<style>` block using the [theme()](https://tailwindcss.com/docs/functions-and-directives#theme) directive, this takes extra wiring, naming another class, etc...

```html
<div class="btn-blue" />
<style>
  .btn-blue {
    --color: theme('colors.blue.500');
  }
</style>
```

---

## Proposal
Expose palette as css variables using one of the following (or mix of)...

```html
<div class="color-blue-500" />
```
Which would set: `--color-blue-500: #3b82f6`

or possibly using an `theme()` and setting a css variable:
```html
<div class="color-[theme(colors.blue.500)]" />
```
Which would set: `--color: #3b82f6`.

or an arbitrary value:
```html
<div class="color-[#123456]" />
```
Which would set: `--color: #123456`.

or setting an specific css variable:
```html
<div class="color-bg-[theme(colors.blue.500)]" />
```
Which would set: `--color-bg: #3b82f6`.

This would allow control of the CSS variable.  You could be more verbose if not supporting arbitrary variables with something like...

```html
<div class="color-blue-500" style="--color: var(--color-blue-500)" />
```

since Tailwind thus far only processes `class` attributes.  If `style` attributes were also processed, it could be something like...

```html
<div style="--color: theme(colors.blue.500)" />
```

**Comment by techniq** [comment]
With the addition of [arbitrary properties](https://github.com/tailwindlabs/tailwindcss/pull/6161), this is now possible with

```html
<div class="[--color:theme(colors.blue.500)]" />
```

**Comment by Olson227** [comment]
code error at runtime .. [PrepaidCardStatus](https://www.prepaidcardstatus.biz/)

**Comment by techniq** [comment]
@Olson227 you need to be using the insider / nightly builds (what will become 3.0)
